gc/base/standard/ConcurrentGCSATB.cpp

```diff
@@ -0,0 +1,298 @@
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup GC_Modron_Standard
+ */
+
+#include "omrcfg.h"
+
+#if defined(OMR_GC_MODRON_CONCURRENT_MARK)
+
+#define J9_EXTERNAL_TO_VM
+
+#include "mmprivatehook.h"
+#include "modronbase.h"
+#include "modronopt.h"
+#include "ModronAssertions.h"
+#include "omr.h"
+
+#include <string.h>
+
+#include "ConcurrentGCSATB.hpp"
+#include "AllocateDescription.hpp"
+
+/**
+ * Create new instance of ConcurrentGCIncrementalUpdate object.
+ *
+ * @return Reference to new MM_ConcurrentGCSATB object or NULL
+ */
+MM_ConcurrentGCSATB *
+MM_ConcurrentGCSATB::newInstance(MM_EnvironmentBase *env)
+{
+	MM_ConcurrentGCSATB *concurrentGC =
+			(MM_ConcurrentGCSATB *)env->getForge()->allocate(sizeof(MM_ConcurrentGCSATB),
+					OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+	if (NULL != concurrentGC) {
+		new(concurrentGC) MM_ConcurrentGCSATB(env);
+		if (!concurrentGC->initialize(env)) {
+			concurrentGC->kill(env);
+			concurrentGC = NULL;
+		}
+	}
+
+	return concurrentGC;
+}
+
+/**
+ * Destroy instance of an ConcurrentGCIncrementalUpdate object.
+ *
+ */
+void
+MM_ConcurrentGCSATB::kill(MM_EnvironmentBase *env)
+{
+	tearDown(env);
+	env->getForge()->free(this);
+}
+
+
+/**
+ * Teardown a MM_ConcurrentGCSATB object
+ * Destroy referenced objects and release
+ * all allocated storage before MM_ConcurrentGCSATB object is freed.
+ */
+void
+MM_ConcurrentGCSATB::tearDown(MM_EnvironmentBase *env)
+{
+	/* ..and then tearDown our super class */
+	MM_ConcurrentGC::tearDown(env);
+}
+
+void
+MM_ConcurrentGCSATB::reportConcurrentHalted(MM_EnvironmentBase *env)
+{}
+
+uintptr_t
+MM_ConcurrentGCSATB::localMark(MM_EnvironmentBase *env, uintptr_t sizeToTrace)
+{
+	omrobjectptr_t objectPtr;
+	uintptr_t gcCount = _extensions->globalGCStats.gcCount;
+
+	env->_workStack.reset(env, _markingScheme->getWorkPackets());
+	Assert_MM_true(env->_cycleState == NULL);
+	Assert_MM_true(CONCURRENT_OFF < _stats.getExecutionMode());
+	Assert_MM_true(_concurrentCycleState._referenceObjectOptions == MM_CycleState::references_default);
+	env->_cycleState = &_concurrentCycleState;
+
+	uintptr_t sizeTraced = 0;
+	while(NULL != (objectPtr = (omrobjectptr_t)env->_workStack.popNoWait(env))) {
+		/* Check for array scanPtr..if we find one ignore it*/
+		if ((uintptr_t)objectPtr & PACKET_ARRAY_SPLIT_TAG){

```

Comment:
Is it correct? If we never split array should it be assert here when we see one? Or, if we do split how we are going to rescan reminder?

Reply:
Ok, this is correct but not optimum. Split array is handled in Work Packet by two sequential records:
- special recored with PACKET_ARRAY_SPLIT_TAG set and array element index we should start at followed by
- standard WP record for indexable object
The split index is ignored here so we start scan at index 0 always (and stop at array end or number of bytes to scan exceeds). The scanning of elements marked already should be relatively fast however does it mean that if number of bytes to scan is large we would have array_size_in_elements/split_amount attempts to rescan this array?